File: typescript/src/secret-projects.ts
import type { HttpClient } from "./http.js";
import type {
  SecretProject,
  SecretMetadata,
  CreateSecretProjectRequest,
  SetProjectSecretRequest,
} from "./types.js";

export class SecretProjects {
  constructor(private http: HttpClient) {}

  create(body: CreateSecretProjectRequest): Promise<SecretProject> {
    return this.http.post(this.http.orgPath("/secret-projects"), body);
  }

  list(): Promise<SecretProject[]> {
    return this.http.get(this.http.orgPath("/secret-projects"));
  }

  delete(projectId: string): Promise<void> {
    return this.http.delete(
      this.http.orgPath(`/secret-projects/${projectId}`),
    );
  }

  async listSecrets(projectId: string): Promise<SecretMetadata[]> {
    const resp = await this.http.get<{ secrets: SecretMetadata[] }>(
      this.http.orgPath(`/secret-projects/${projectId}/secrets`),
    );
    return resp.secrets ?? [];
  }

  setSecret(projectId: string, body: SetProjectSecretRequest): Promise<void> {
    return this.http.put(
      this.http.orgPath(`/secret-projects/${projectId}/secrets`),
      body,
    );
  }

  deleteSecret(projectId: string, key: string): Promise<void> {
    return this.http.delete(
      this.http.orgPath(
        `/secret-projects/${projectId}/secrets/${encodeURIComponent(key)}`,
      ),
    );
  }
}

File: go/secret_projects.go
package codewire

import (
	"context"
	"net/url"

	"codewire.sh/sdk-go/internal/api"
)

type SecretProjectService struct {
	client *Client
}

func (s *SecretProjectService) Create(ctx context.Context, req api.CreateSecretProjectInputBody) (*api.SecretProject, error) {
	path, err := s.client.orgPath("/secret-projects")
	if err != nil {
		return nil, err
	}
	var proj api.SecretProject
	if err := s.client.do(ctx, "POST", path, req, &proj); err != nil {
		return nil, err
	}
	return &proj, nil
}

func (s *SecretProjectService) List(ctx context.Context) ([]api.SecretProject, error) {
	path, err := s.client.orgPath("/secret-projects")
	if err != nil {
		return nil, err
	}
	var projects []api.SecretProject
	if err := s.client.do(ctx, "GET", path, nil, &projects); err != nil {
		return nil, err
	}
	return projects, nil
}

func (s *SecretProjectService) Delete(ctx context.Context, projectID string) error {
	path, err := s.client.orgPath("/secret-projects/%s", projectID)
	if err != nil {
		return err
	}
	return s.client.do(ctx, "DELETE", path, nil, nil)
}

func (s *SecretProjectService) ListSecrets(ctx context.Context, projectID string) ([]api.SecretMetadata, error) {
	path, err := s.client.orgPath("/secret-projects/%s/secrets", projectID)
	if err != nil {
		return nil, err
	}
	var resp api.ListProjectSecretsOutputBody
	if err := s.client.do(ctx, "GET", path, nil, &resp); err != nil {
		return nil, err
	}
	if resp.Secrets == nil {
		return nil, nil
	}
	return *resp.Secrets, nil
}

func (s *SecretProjectService) SetSecret(ctx context.Context, projectID string, key, value string) error {
	path, err := s.client.orgPath("/secret-projects/%s/secrets", projectID)
	if err != nil {
		return err
	}
	req := api.SetProjectSecretInputBody{
		Key:   key,
		Value: value,
	}
	return s.client.do(ctx, "PUT", path, req, nil)
}

func (s *SecretProjectService) DeleteSecret(ctx context.Context, projectID, key string) error {
	path, err := s.client.orgPath("/secret-projects/%s/secrets/%s", projectID, url.PathEscape(key))
	if err != nil {
		return err
	}
	return s.client.do(ctx, "DELETE", path, nil, nil)
}
